File: integration/ts/p2pkh.test.ts
/**
 * P2PKH integration test — stateless contract with checkSig.
 *
 * P2PKH locks funds to a public key hash. Spending requires a valid
 * signature and the matching public key. The SDK auto-computes Sig params
 * when null is passed.
 */

import { describe, it, expect } from 'vitest';
import { compileContract } from './helpers/compile.js';
import { RunarContract } from 'runar-sdk';
import { createFundedWallet, createWallet } from './helpers/wallet.js';
import { createProvider } from './helpers/node.js';

describe('P2PKH', () => {
  it('should compile and deploy with a valid pubKeyHash', async () => {
    const artifact = compileContract('examples/ts/p2pkh/P2PKH.runar.ts');
    expect(artifact).toBeTruthy();
    expect(artifact.contractName).toBe('P2PKH');

    const provider = createProvider();
    const { signer, pubKeyHash } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [pubKeyHash]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 5000 });
    expect(deployTxid).toBeTruthy();
    expect(typeof deployTxid).toBe('string');
    expect(deployTxid.length).toBe(64);
  });

  it('should deploy and spend with unlock(sig, pubKey)', async () => {
    const artifact = compileContract('examples/ts/p2pkh/P2PKH.runar.ts');

    const provider = createProvider();
    const { signer, pubKeyHash } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [pubKeyHash]);
    await contract.deploy(provider, signer, { satoshis: 5000 });

    // null Sig and PubKey args are auto-computed by the SDK
    const { txid: callTxid } = await contract.call(
      'unlock', [null, null], provider, signer,
    );
    expect(callTxid).toBeTruthy();
    expect(callTxid.length).toBe(64);
  });

  it('should deploy with a different pubKeyHash', async () => {
    const artifact = compileContract('examples/ts/p2pkh/P2PKH.runar.ts');

    const provider = createProvider();
    const { signer } = await createFundedWallet(provider);

    const otherWallet = createWallet();
    const contract = new RunarContract(artifact, [otherWallet.pubKeyHash]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 5000 });
    expect(deployTxid).toBeTruthy();
  });

  it('should reject unlock with wrong signer', async () => {
    const artifact = compileContract('examples/ts/p2pkh/P2PKH.runar.ts');

    const provider = createProvider();
    // Deploy locked to walletA's pubKeyHash
    const walletA = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [walletA.pubKeyHash]);
    await contract.deploy(provider, walletA.signer, { satoshis: 5000 });

    // Call unlock with walletB — auto-computed sig+pubkey won't match the hash
    const walletB = await createFundedWallet(provider);

    await expect(
      contract.call('unlock', [null, null], provider, walletB.signer),
    ).rejects.toThrow();
  });
});

File: integration/zig/src/p2pkh_test.zig
const std = @import("std");
const runar = @import("runar");
const bsvz = @import("bsvz");
const helpers = @import("helpers.zig");
const compile = @import("compile.zig");

/// Deploy a P2PKH contract locked to the given owner wallet.
fn deployP2PKH(allocator: std.mem.Allocator, owner: *helpers.Wallet) !struct { contract: runar.RunarContract, provider: helpers.RPCProvider, signer: runar.LocalSigner } {
    var artifact = try compile.compileContract(allocator, "examples/ts/p2pkh/P2PKH.runar.ts");
    errdefer artifact.deinit();

    const pkh_hex = try owner.pubKeyHashHex(allocator);
    defer allocator.free(pkh_hex);

    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = pkh_hex },
    });
    errdefer contract.deinit();

    // Fund the owner wallet
    const fund_txid = try helpers.fundWallet(allocator, owner, 1.0);
    allocator.free(fund_txid);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var local_signer = try owner.localSigner();

    // Deploy
    const txid = try contract.deploy(rpc_provider.provider(), local_signer.signer(), .{ .satoshis = 5000 });
    defer allocator.free(txid);
    std.log.info("P2PKH deployed: {s}", .{txid});

    return .{ .contract = contract, .provider = rpc_provider, .signer = local_signer };
}

test "P2PKH_ValidUnlock" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var owner = try helpers.newWallet(allocator);
    defer owner.deinit();

    var artifact = compile.compileContract(allocator, "examples/ts/p2pkh/P2PKH.runar.ts") catch |err| {
        std.log.warn("Could not compile P2PKH contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    std.log.info("P2PKH script: {d} bytes", .{artifact.script.len / 2});

    const pkh_hex = try owner.pubKeyHashHex(allocator);
    defer allocator.free(pkh_hex);

    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = pkh_hex },
    });
    defer contract.deinit();

    // Fund
    const fund_txid = try helpers.fundWallet(allocator, &owner, 1.0);
    defer allocator.free(fund_txid);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var local_signer = try owner.localSigner();

    // Deploy
    const deploy_txid = try contract.deploy(rpc_provider.provider(), local_signer.signer(), .{ .satoshis = 5000 });
    defer allocator.free(deploy_txid);
    std.log.info("P2PKH deployed: {s}", .{deploy_txid});

    try std.testing.expectEqual(@as(usize, 64), deploy_txid.len);

    // The P2PKH contract is stateless and requires signature verification.
    // For a full unlock, we would need to build a spending tx and compute
    // the BIP-143 sighash. For now, verify the deployment succeeded.
    const utxo = contract.getCurrentUtxo();
    try std.testing.expect(utxo != null);
    try std.testing.expectEqual(@as(i64, 5000), utxo.?.satoshis);
}

test "P2PKH_DeployDifferentPubKeyHash" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/ts/p2pkh/P2PKH.runar.ts") catch |err| {
        std.log.warn("Could not compile P2PKH contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    // Owner 1
    var owner1 = try helpers.newWallet(allocator);
    defer owner1.deinit();
    const pkh1 = try owner1.pubKeyHashHex(allocator);
    defer allocator.free(pkh1);

    var contract1 = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = pkh1 },
    });
    defer contract1.deinit();

    const fund1 = try helpers.fundWallet(allocator, &owner1, 1.0);
    defer allocator.free(fund1);

    var rpc1 = helpers.RPCProvider.init(allocator);
    var signer1 = try owner1.localSigner();
    const txid1 = try contract1.deploy(rpc1.provider(), signer1.signer(), .{ .satoshis = 5000 });
    defer allocator.free(txid1);

    // Owner 2
    var owner2 = try helpers.newWallet(allocator);
    defer owner2.deinit();
    const pkh2 = try owner2.pubKeyHashHex(allocator);
    defer allocator.free(pkh2);

    var contract2 = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = pkh2 },
    });
    defer contract2.deinit();

    const fund2 = try helpers.fundWallet(allocator, &owner2, 1.0);
    defer allocator.free(fund2);

    var rpc2 = helpers.RPCProvider.init(allocator);
    var signer2 = try owner2.localSigner();
    const txid2 = try contract2.deploy(rpc2.provider(), signer2.signer(), .{ .satoshis = 5000 });
    defer allocator.free(txid2);

    // Verify different txids (different pubkey hashes produce different scripts)
    try std.testing.expect(!std.mem.eql(u8, txid1, txid2));
    std.log.info("owner1 txid: {s}, owner2 txid: {s}", .{ txid1, txid2 });
}
